config/base/derive/src/lib.rs

```diff
@@ -0,0 +1,685 @@
+                            LocalError::ConflictNested(span) => Self::new(span, "attributes conflict: `nested` cannot be set with other attributes"),
+                            LocalError::ConflictEnv(span) => Self::new(span, "attributes conflict: set either `env` or `env_custom`, not both"),
+                            LocalError::Duplicate(span) => Self::new(span, "duplicate attribute"),
+                            LocalError::BadDefaultExpr(span, error) => Self::new(span, format!("expected a valid expression within `default = \"<expr>\"`, but couldn't parse it: {error}")),
+                            LocalError::EnvExpectedFormat(span) => Self::new(span, "expected `env` to be set as `env = \"VARIABLE_NAME\""),
+                            LocalError::EnvOnlyWithoutEnv(span) => Self::new(span, "`env_only` cannot be set without `env`"),
+                            LocalError::UnexpectedToken(span) => Self::new(span, "unexpected token; expected `default`, `env`, `env_only`, or `nested`"),
+                        }
+                    }
+                }
+
+                let mut attr_default = None;
+                let mut attr_env = None;
+                let mut attr_env_custom = false;
+                let mut attr_nested = false;
+
+                let mut rest = *cursor;
+
+                while let Some((tt, next)) = rest.token_tree() {
+                    match &tt {
+                        TokenTree::Ident(ident) => {
+                            let token = ident.to_string();
+                            match token.as_str() {
+                                "default" => {
+                                    if attr_nested  {
+                                        Err(LocalError::ConflictNested(ident.span()))?
+                                    }
+                                    if attr_default.is_some() {
+                                        Err(LocalError::Duplicate(ident.span()))?
+                                    }
+
+                                    let (lit, next) = match expect_comma_or_eq_with_lit_str(next)? {
+                                        (None, next) => {
+                                            attr_default = Some(AttrDefault::Word);
+                                            rest = next;
+                                            continue;
+                                        }
+                                        (Some(lit), next) => {
+                                            (lit, next)
+                                        },
+                                    };
+                                    let expr: syn::Expr = lit
+                                        .parse()
+                                        .map_err(|err| LocalError::BadDefaultExpr(lit.span(), err))?;
+                                    attr_default = Some(AttrDefault::Expr(expr));
+                                    rest = expect_comma_or_none(next)?;
```

Comment:
Is there a particular reason you decided agains re-using syn's `syn::punctuated::Punctuated::<YourTokenType, Token![,]>` for parsing commans and decided to roll out your own comma parser?

One unfortunate thing is that you have to add a separate type for a token and then implement `Parse` for it, but as a bonus you can have commas (and optional trailing ones) handled for you

Reply:
No reason apart from my ignorance of its existence!

Will definitely try to re-use it.